src/main/java/be/seeseemelk/mockbukkit/scheduler/BukkitSchedulerMock.java

```diff
@@ -281,67 +304,91 @@ public boolean isQueued(int taskId)
 	{
 		// TODO Auto-generated method stub
 		throw new UnimplementedOperationException();
 	}
 
 	@Override
-	public List<BukkitTask> getPendingTasks()
+	public @NotNull List<BukkitTask> getPendingTasks()
 	{
 		// TODO Auto-generated method stub
 		throw new UnimplementedOperationException();
 	}
 
 	@Override
-	public BukkitTask runTaskAsynchronously(Plugin plugin, Runnable task)
+	public @NotNull BukkitTask runTaskAsynchronously(@NotNull Plugin plugin, @NotNull Runnable task)
 	{
+
 		ScheduledTask scheduledTask = new ScheduledTask(id++, plugin, false, currentTick, new AsyncRunnable(task));
+		checkForShutDown(scheduledTask);
 		asyncTasksRunning.incrementAndGet();
 		pool.execute(scheduledTask.getRunnable());
 		return scheduledTask;
 	}
 
 	@Override
-	public BukkitTask runTaskAsynchronously(Plugin plugin, BukkitRunnable task)
+	public @NotNull BukkitTask runTaskAsynchronously(@NotNull Plugin plugin, @NotNull BukkitRunnable task)
 	{
 		return runTaskAsynchronously(plugin, (Runnable) task);
 	}
 
 	@Override
-	public BukkitTask runTaskLater(Plugin plugin, BukkitRunnable task, long delay)
+	public @NotNull BukkitTask runTaskLater(@NotNull Plugin plugin, @NotNull BukkitRunnable task, long delay)
 	{
 		return runTaskLater(plugin, (Runnable) task, delay);
 	}
 
 	@Override
-	public BukkitTask runTaskLaterAsynchronously(Plugin plugin, Runnable task, long delay)
+	public @NotNull BukkitTask runTaskLaterAsynchronously(@NotNull Plugin plugin, @NotNull Runnable task, long delay)
 	{
 		ScheduledTask scheduledTask = new ScheduledTask(id++, plugin, false, currentTick + delay,
 		        new AsyncRunnable(task));
-		tasks.add(scheduledTask);
+		checkForShutDown(scheduledTask);
 		asyncTasksQueued++;
+		tasks.add(scheduledTask);
 		return scheduledTask;
 	}
 
 	@Override
-	public BukkitTask runTaskLaterAsynchronously(Plugin plugin, BukkitRunnable task, long delay)
+	public @NotNull BukkitTask runTaskLaterAsynchronously(@NotNull Plugin plugin, @NotNull BukkitRunnable task, long delay)
 	{
 		return runTaskLaterAsynchronously(plugin, (Runnable) task, delay);
 	}
 
 	@Override
-	public BukkitTask runTaskTimerAsynchronously(Plugin plugin, Runnable task, long delay, long period)
+	public @NotNull BukkitTask runTaskTimerAsynchronously(@NotNull Plugin plugin, @NotNull Runnable task, long delay, long period)
 	{
 		RepeatingTask scheduledTask = new RepeatingTask(id++, plugin, false, currentTick + delay, period,
 		        new AsyncRunnable(task));
+		checkForShutDown(scheduledTask);
 		tasks.add(scheduledTask);
 		return scheduledTask;
 	}
 
+	private void checkForShutDown(ScheduledTask task)
+	{
+		if (shuttingDown)
+		{
+			String name;
+			if (task.getOwner() != null)
+			{
+				name = task.getOwner().getName();
+			}
+			else
+			{
+				name = "";
+			}
+			Logger.getLogger(LOGGER_NAME).warning(
+			    name + "(" + task.getRunnable().getClass().getSimpleName()
+			    +  "): Scheduler is shutting down - why are you scheduling tasks - CANCELLED");
```

Comment:
I would prefer to make this an exception. A unit test should be correct. We shouldn't attempt to run code that is _somewhat_ correct, given that we are testing if code _is_ correct..

Reply:
Bukkit does exactly the same thing if you cause a task to run after shutdown - its bad practice but not breaking code

Reply:
ie it doesnt fail but it doesnt schedule